cudax/include/cuda/experimental/detail/dimensions.cuh

```diff
@@ -0,0 +1,141 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX_DETAIL_DIMENSIONS
+#define _CUDAX_DETAIL_DIMENSIONS
+
+#include <cuda/std/mdspan>
+
+namespace cuda::experimental
+{
+
+template <typename T, size_t... Extents>
+using dimensions = ::cuda::std::extents<T, Extents...>;
+
+// not unsigned because of a bug in ::cuda::std::extents
+using dimensions_index_type = int;
+
+/**
+ * @brief Type representing a result of a multi-dimensional hierarchy query.
+ *
+ * Returned from extents and index queries.
+ *
+ * @par Snippet
+ * @code
+ * #include <cudax/hierarchy_dimensions.cuh>
+ *
+ * template <typename Dimensions>
+ * __global__ void kernel(Dimensions dims)
+ * {
+ *     auto ext = dims.extents();
+ *
+ *     // Can be accessed like cuda::std::extents or like dim3
+ *     assert(ext.extent(0) == expected);
+ *     assert(ext.x == expected);
+ *
+ *     // Can be converted to dim3
+ *     dim3 dimensions = ext;
+ * }
+ * @endcode
+ * @par
+ *
+ * @tparam T
+ *   Type of the result for each dimension
+ *
+ * @tparam Extents
+ *   Extents of the result
+ */
+template <typename T, size_t... Extents>
+struct hierarchy_query_result : public dimensions<T, Extents...>
+{
+  using Dims = dimensions<T, Extents...>;
+  using Dims::Dims;
+  _CCCL_HOST_DEVICE explicit constexpr hierarchy_query_result(const Dims& dims)
+      : Dims(dims)
+  {}
+  static_assert(Dims::rank() > 0 && Dims::rank() <= 3);
+
+  const T x = Dims::extent(0);
+  const T y = Dims::rank() > 1 ? Dims::extent(1) : 1;
+  const T z = Dims::rank() > 2 ? Dims::extent(2) : 1;
+
+  constexpr _CCCL_HOST_DEVICE operator dim3() const
+  {
+    return dim3(static_cast<uint32_t>(x), static_cast<uint32_t>(y), static_cast<uint32_t>(z));
+  }
+};
+
+namespace detail
+{
+template <typename OpType>
+_CCCL_HOST_DEVICE constexpr size_t merge_extents(size_t e1, size_t e2)
```

Comment:
General question. Should this be:
```suggestion
_CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr size_t merge_extents(size_t e1, size_t e2)
```

Applies throughout

Reply:
Added _CCCL_NODISCARD to all functions in detail namespace, but I had an issue where on a few functions compiler complained about it being applied to a function returning void. There is no way these functions return void, so I just left the annotation commented out for now